community/rfcs/techniques/design/20200225-keras-pruning-design.md

```diff
@@ -0,0 +1,539 @@
+
+*   Do not prune very frequently to give the model time to recover. The
+    [pruning schedule](https://www.tensorflow.org/model_optimization/api_docs/python/tfmot/sparsity/keras/PruningSchedule)
+    provides a decent default frequency.
+
+*   For general ideas to improve model accuracy, find your use case(s) under
+    "Define model" on the navigation sidebar and see if there are tips.
+
+### Checkpointing and deserialization
+
+**Your Use Case:**
+
+*   You cannot do checkpointing with Keras HDF5 weights since we need to
+    preserve the step.
+
+*   This code is only needed for the HDF5 model format (not HDF5 weights or
+    other formats).
+
+```
+model = setup_model()
+pruned_model = tfmot.sparsity.keras.prune_low_magnitude(model)
+
+_, keras_model_file = tempfile.mkstemp('.h5')
+# Saving the optimizer is necessary for checkpointing (True is the default).
+pruned_model.save(keras_model_file, include_optimizer=True)
+
+with tfmot.sparsity.keras.prune_scope():
+  loaded_model = tf.keras.models.load_model(keras_model_file)
+
+```
+
+### Deployment
+
+#### Export model with size compression
+
+**Common mistake**: both `strip_pruning` and applying a standard compression
+algorithm (e.g. via gzip) are necessary to see the compression benefits of
+pruning.
+
+```
+# See "Define model" and "Train model" on navigation sidebar for how to define
+# and train this model in other ways.
+model = setup_model()
+pruned_model = tfmot.sparsity.keras.prune_low_magnitude(model)
+
+pruned_model.compile(
+      loss=tf.keras.losses.categorical_crossentropy,
+      optimizer='adam',
+      metrics=['accuracy']
+)
+
+pruned_model.fit(
+    x_train,
+    y_train,
+    callbacks=[tfmot.sparsity.keras.UpdatePruningStep()]
+)
+
+final_model = tfmot.sparsity.keras.strip_pruning(pruned_model)
+```
+
+#### Hardware-specific optimizations
+
+Once the framework
+[enables pruning to improve latency](\(https://github.com/tensorflow/model-optimization/issues/173\)),
+using block sparsity can improve latency for certain hardware. For a target
+model accuracy, latency can still improve despite the fact that increasing the
+block size will decrease the peak sparsity %.
+
+```
+model = setup_model()
+
+# For using intrinsics on a CPU with 128-bit registers, together with 8-bit
+# quantized weights, a 1x16 block size is nice because the block perfectly
+# fits into the register.
+pruning_params = {'block_size': [1, 16]}
+pruned_model = tfmot.sparsity.keras.prune_low_magnitude(model, **pruning_params)
+```
+
+<!-- See the technique overview page documentation template for what to consider
+here. Includes potential support matrix section. -->
+
+## Compatibility
+
+### At launch
+
+Users can apply pruning with the following APIs:
+
+*   Model building: `tf.keras` with only Sequential and Functional models
+*   TensorFlow versions: TF 1.x for versions 1.14+ and 2.x.
+    *   `tf.compat.v1` with a TF 2.X package and `tf.compat.v2` with a TF 1.X
```

Comment:
This statement is generally true. The comment in the code provides partial context on why compat.v1 and compat.v2 is done specifically for RNN_CELL_WEIGHTS_MAP. In particular, 
the TF 2.X API's RNN actually used the TF 1.X API's RNNCell. This was something that I pointed out and should later have been fixed, but we need to keep it for backwards compatibility for older TF versions. 

For the quantization-aware tool that's launching, we may be able to skip it since it will require tf-nightly, which should no longer after the above issue.

Reply:
|  In particular, the TF 2.X API's RNN actually used the TF 1.X API's RNNCell.

So when I first tried just using tf.keras.layers.RNNCell (or the equivalent) it threw an error when I was using a TF 2.X package saying that the TF 1.X RNNCell was not whitelisted.